docs/Development.md

```diff
@@ -40,30 +40,80 @@ More on pyfpdf:
+auto-format source code before any commit.
+
+## Linting ##
+
+We use [pylint](https://github.com/PyCQA/pylint/) as a static code analyzer
+to detect potential issues in the code.
+
+In case of special "false positive" cases,
+checks can be disabled locally with `#pylint disable=XXX` code comments,
+or globally through the `.pylintrc` file.
+
+## Pre-commit hook ##
+If you use a UNIX system, you can place the following shell code
+in `.git/hooks/pre-commit` in order to always invoke `black` & `pylint`
+before every commit:
+
+```shell
+#!/bin/sh
+modified_py_files=$(git diff --cached --name-only --diff-filter=ACM | grep '.py$')
+if [ -n "$modified_py_files" ]; then
+    if ! black --check $modified_py_files; then
+        black $modified_py_files
+        exit 1
+    fi
+    pylint $modified_py_files
+fi
+```
+
+It will abort the commit if `pylint` found issues
+or `black` detect non-properly formatted code.
+In the later case though, it will auto-format your code
+and you will just have to run `git commit -a` again.
 
 ## Testing ##
 
-Testing is done with [Tox](https://tox.readthedocs.io/en/latest/), and is
-self-documented in the `tox.ini` file in the repository. To run tests, cd into
-the cloned repository and run `tox`.
+To run tests, `cd` into the repository and run `python setup.py test`.
 
-If you do not want to run tests for all versions of python, run `tox -e py27`
-(or your version of python). To install all versions of python that are
-supported on Ubuntu, see the instructions on the Github Repository home page
-of this project.
+To only run tests for your version of Python, run `python setup.py test`.
 
-Be sure to see the example tests in the `test` folder & `test\utilities.py`
-and explore that folder in general.
+When a unit test generates a PDF, it is recommended to use the `assert_pdf_equal`
+utility function in order to validate the output.
+It relies on the very handy [qpdf](https://github.com/qpdf/qpdf) CLI program
+to generate a PDF that is easy to compare: annotated, strictly formatted,
+whith uncompressed internal streams.
+You will need to have its binary in your `$PATH`,
+otherwise `assert_pdf_equal` will fall back to hash-based comparison.
+In order to generate a "reference" PDF file, simply call `assert_pdf_equal` once
+with `delete=False`, and retrieve the generated `/tmp/pyfpdf-test-*-actual.pdf` file.
 
-## Documentation ##
+Be sure to see the example tests in the `test/` folder in general.
+
+## GitHub pipeline ##
 
-To build docs, cd into repository and `tox -e docs`.
+A [GitHub Actions](https://help.github.com/en/actions/reference) pipeline
+is executed on every commit on the `master` branch,
+and for every _Pull Request_.
+
+It performs all validation steps detailed above: code checking with `black`,
+linting with `pylint`, unit tests...
+_Pull Requests_ submitted must pass all those checks in order to be approved.
+Ask maintainers through comments if some errors in the pipeline seem obscure to you.
```

Comment:
tox was also useful for this as well, less for contributors to do when trying to preview changes (in this case, to docs).

Reply:
I agree it was, but with the new CI pipeline this is not really useful anymore:
GitHub Actions provide a matrix of pre-installed Python versions.

Reply:
I would like to preserve that so it is able to be run locally

Reply:
ok